src/java.base/share/classes/jdk/internal/vm/VMSupport.java

```diff
@@ -167,4 +179,404 @@ public static int encodeThrowable(Throwable throwable, long buffer, int bufferSi
         return requiredSize;
     }
+
+    /**
+     * Parses {@code rawAnnotationBytes} into a list of {@link Annotation}s and then
+     * serializes them to a byte array with {@link #encodeAnnotations(Collection)}.
+     */
+    public static byte[] encodeAnnotations(byte[] rawAnnotations,
+                                           Class<?> declaringClass,
+                                           ConstantPool cp,
+                                           boolean forClass,
+                                           Class<? extends Annotation>[] selectAnnotationClasses)
+    {
+        for (Class<?> c : selectAnnotationClasses) {
+            if (!c.isAnnotation()) {
+                throw new IllegalArgumentException(c + " is not an annotation interface");
+            }
+        }
+        Map<Class<? extends Annotation>, Annotation> annotations =
+                AnnotationParser.parseSelectAnnotations(rawAnnotations, cp, declaringClass, selectAnnotationClasses);
+        if (forClass && annotations.size() != selectAnnotationClasses.length) {
+            Class<?> superClass = declaringClass.getSuperclass();
+            nextSuperClass:
+            while (superClass != null) {
+                JavaLangAccess jla = SharedSecrets.getJavaLangAccess();
+                Map<Class<? extends Annotation>, Annotation> superAnnotations =
+                    AnnotationParser.parseSelectAnnotations(
+                            jla.getRawClassAnnotations(superClass),
+                            jla.getConstantPool(superClass),
+                            superClass,
+                            selectAnnotationClasses);
+
+                for (Map.Entry<Class<? extends Annotation>, Annotation> e : superAnnotations.entrySet()) {
+                    Class<? extends Annotation> annotationClass = e.getKey();
+                    if (!annotations.containsKey(annotationClass) && AnnotationType.getInstance(annotationClass).isInherited()) {
+                        if (annotations.isEmpty()) {
+                            // An empty map might be unmodifiable (e.g. Collections.emptyMap()).
+                            annotations = new LinkedHashMap<Class<? extends Annotation>, Annotation>();
+                        }
+                        annotations.put(annotationClass, e.getValue());
+                        if (annotations.size() == selectAnnotationClasses.length) {
+                            break nextSuperClass;
+                        }
+                    }
+                }
+                superClass = superClass.getSuperclass();
+            }
+        }
+        return encodeAnnotations(annotations.values());
+    }
+
+    /**
+     * Encodes a list of annotations to a byte array. The byte array can be decoded with {@link #decodeAnnotations(byte[], AnnotationDecoder)}.
+     */
+    public static byte[] encodeAnnotations(Collection<Annotation> annotations) {
```

Comment:
I don't think it matters much in this use case, but it looks like  encodeAnnotations could be changed to take a List rather than a Collection, as the comment implies.

Reply:
Just above (line 228) you can see a call to this method where the argument comes from `Map.values()` whose type is `Collection` so I'd prefer to leave it as is rather than have to convert the argument to a `List`.

Reply:
Just above (line 228) you can see a call to this method where the argument comes from `Map.values()` whose type is `Collection` so I'd prefer to leave it as is rather than have to convert the argument to a `List`.

Reply:
In that case, I think the comment on line 232 should be updated to not say "list".